# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
# lambda parameters: expression

print(lambda num: num + 1)

print((lambda num: num + 1)(4))

<function <lambda> at 0x000001AC0BC824C0>
5


#### Saving to a Variable

In [3]:
### regular function ###
# def addone(num):
#     return num + 1


### lambda function ###
addone = lambda num: num + 1




print(addone(4))

5


#### Multiple Inputs

In [4]:
find_vol = lambda length,width,height: length * width * height

print(find_vol(2,2,3))

12


#### Passing a Lambda into a Function

In [5]:
def multiply(function, num):
    return function(num)                            # pass 4 into lambda

print(multiply(lambda num: num * num, 4))

16


#### Returning a Lambda from a Function

In [7]:
def set_exponent(num):
    return lambda l_num: l_num ** num

square = set_exponent(2) #lambda l_num: l_num ** 2

print(square(4))                                        #lambda l_num: 4 ** 2

cube = set_exponent(3) #lambda l_num: l_num ** 3

print(cube(5))                                          #lambda l_num: 5 ** 3

16
125


#### If Statements within Lambdas

In [9]:
# Lambda x: Do this if condition is True else (False) Do that
is_even = lambda num: f'{num} is even' if num % 2 == 0 else f'{num} is odd'

print(is_even(4))

4 is even


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [1]:
f = lambda num: num ** 3

print(f(2))

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [3]:
##           map(function, iterable(string, list, dictionary, set))          ##


### regular function ###
my_list = [12,45,2,34]

def double_nums(alist):
    doubled = []
    for num in alist:
        doubled.append(num * 2)
    return doubled


print(double_nums(my_list))


### map function ###
def map_double(element):
    return element * 2

#          map(function_name, list_name)

print(list(map(map_double, my_list)))

[24, 90, 4, 68]
[24, 90, 4, 68]


#### Using Lambda's with Map

In [16]:
#          map(lambda function, list_name)

print(list(map(lambda num: num * 2, my_list)))

[24, 90, 4, 68]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [4]:
num_list = [2,6,99,2,45,56]


#           map(lambda function, list_name)

print(list(map(lambda num: (num * 2) - 1, num_list)))

[3, 11, 197, 3, 89, 111]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [5]:
### Regular Function ###
def only_odds(alist):
    odd_nums = []
    for num in alist:
        if num % 2 != 0:
            odd_nums.append(num)
    return odd_nums

new_list = [1,12,54,41,23456,23]

print(only_odds(new_list))


# True if (condition) else False

#return True if num %2 != 0 else False 

#return num %2 != 0



### Filter Function ###
def filter_odds(num):
    return num %2 != 0                  # return True if num is odd             (the filter function then returns the values that passed the condition, num %2 != 0)

print(list(filter(filter_odds, new_list)))                                      # filter([True, False, False, True, False, True], [1, 12, 54, 41, 23456, 23])



                                                                                # The filter function essentially uses a function, filter_odds, to see if each element in a list (or any iterable type), new_list, passes a condition
                                                                                # (returns True). Once the function has returned a bunch of True's, it then matches up each True with its corresponding element (matching indexes).
                                                                                # So basically, a filter function is compares two lists: one list of booleans and one list provided by the programmer.


[1, 41, 23]
[1, 41, 23]


In [7]:
def filter_odds(num):
    return num %2 != 0                  # return True if num is odd             (the filter function then returns the values that passed the condition, num %2 != 0)

print(list(map(filter_odds, new_list)))                                      # filter(True, False, False, True, False, True)


[True, False, False, True, False, True]


#### Using Lambda's with Filter()

In [23]:
                                                                                    ################ RESUME READING HERE ################


# Lambda x: True if (condition) else False

print(list(filter(lambda num: num %2 != 0 , new_list)))

#How python reads this expression
print(2 % 2 != 0)



[1, 41, 23]
False


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [ ]:
# [2,6,99,2,45,56]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [26]:
from functools import reduce

new_list = [1,2,3,4,5,6]

def add_nums(alist):
    total = 0 
    for num in alist:
        total += num
    return total

print(add_nums(new_list))
    


# reduce(function, iterable)

def reduce_add(num1, num2):
    print(num1, num2)
    return num1 * num2

print(reduce(reduce_add, new_list))



21
1 2
2 3
6 4
24 5
120 6
720


#### Using Lambda's with Reduce()

In [27]:
reduce(lambda num1, num2: num1 + num2, new_list)

21

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [28]:
reduce(lambda num1, num2: num1 * num2, new_list)

720

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [31]:
def add_nums(num):
    # 5 + 4 + 3 + 2 + 1 = 15
    #base case
    if num <=1:
        print(f'add_nums({num}) = {num}')
        return num
    else:
        #recursive case
        print(f'add_nums({num}) = {num} + add_nums({num-1})')
        return num * add_nums(num - 1)
    
print(add_nums(5))

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1
120


#### Writing a Factorial Function

In [33]:
#5 * 4 *3 * 2 *1

#good easy usecase for recursion

def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num-1)
    
print(factorial(5))

120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [3]:
def sqr_gen(num):
    while num > 0:
        yield num ** 2
        num -= 1

for i in sqr_gen(10):
    print(i)


100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [10]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def filter_cities(element):
    return element.strip()

print(list(filter(filter_cities, places)))

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [11]:
import re

authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# https://www.w3schools.com/python/ref_list_sort.asp

#build a key that will hone in on the last name of each person




def get_last_name(full_name):
    last_name = full_name.split()[-1]                   # ' ' is default value for split()
    
    return last_name.lower()



sort_last_names = lambda aList: aList.sort(key=get_last_name)

sort_last_names(authors)

print(authors)



['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [14]:
# F = (9/5)*C + 32


places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

#                                          return a tuple as each element in list
updated_temps = list(map(lambda city_temp: (city_temp[0], (9/5)*city_temp[1] + 32), places))
print(updated_temps)



[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [15]:

def fib(n):
    # base case
    if n == 1 or n == 2:
        return 1
    # recursive case
    else:
        return (fib(n-1) + fib(n-2))


for num in range(1, 7):
    print(f"Iteration {num-1}: {fib(num)}")


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
